### PR TITLE
[SPARK-45360][SQL][CONNECT] Initialize spark session builder configuration from SPARK_REMOTE

### PR DESCRIPTION
**Pull request by heyihong**
### What changes were proposed in this pull request?

- Initialize spark session builder configuration from SPARK_REMOTE

### Why are the changes needed?

- `SparkSession.builder().getOrCreate()` should follow the behavior documents [here](https://github.com/apache/spark/blob/2cc1ee4d3a05a641d7a245f015ef824d8f7bae8b/docs/spark-connect-overview.md?plain=1#L241-L244) and support initialization from SPARK_REMOTE

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

- `build/sbt "connect-client-jvm/testOnly *SparkConnectClientSuite"`

### CODE REVIEW

**Comment by heyihong** [comment]
@hvanhovell @juliuszsompolski Please take a look

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -912,6 +912,11 @@ object SparkSession extends Logging {
     @deprecated(message = "Please use create() instead.", since = "3.5.0")
     def build(): SparkSession = create()
 
+    private def getClientConfiguration(): Configuration = {
+      // Use copy() to avoid modifying the state of the existing builder.
+      builder.copy().loadFromEnvironment().configuration
```
Comment (juliuszsompolski) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:917]: in this case, SPARK_REMOTE will take precedence over what was explicitly set in the builder, right?
so having in env
```
SPARK_REMOTE="sc://foo:123"
```
and calling
```
SparkSession.builder.remote("sc://bar:234").create()
```
then the SPARK_REMOTE will take precedence? Documentation does not specify precedence, but it feels like setting `remote` explicitly should take precedence over env var (and it should also be specified in the documentation)

Comment (juliuszsompolski) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:917]: In python, the explicity set remote config set in https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/session.py#L143 takes precedence over env in https://github.com/apache/spark/blob/master/python/pyspark/sql/session.py#L470

It should be the same in scala, and the precedence should be documented.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -923,7 +928,7 @@ object SparkSession extends Logging {
      */
     def create(): SparkSession = {
       val session = tryCreateSessionFromClient()
-        .getOrElse(SparkSession.this.create(builder.configuration))
+        .getOrElse(SparkSession.this.create(getClientConfiguration()))
       setDefaultAndActiveSession(session)
       applyOptions(session)
       session
```
Comment (juliuszsompolski) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:934]: Orthogonal: Why are create and getOrCreate functions exactly the same, while I think they should be doing something slightly different??

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
Comment (heyihong) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:917]: SPARK_REMOTE value makes the final override before [RemoteSparkSession.builder.config(map=opts).getOrCreate()](https://github.com/apache/spark/blob/master/python/pyspark/sql/session.py#L479) if I understand correctly?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 2 above)
Comment (heyihong) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:934]: `getOrCreate()` will try to retrieve cached SparkSession if exists. `create()` only creates a new SparkSession

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala [hunk 3]
```diff
@@ -112,6 +130,35 @@ class SparkConnectClientSuite extends ConnectFunSuite with BeforeAndAfterEach {
     }
   }
 
+  test("SparkSession create with SPARK_REMOTE") {
```
Comment (vsevolodstep-db) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala:133]: It's worth to also have a test covering combination of both `SPARK_REMOTE` env variable and `builder.remote(...)` call

**Review by vsevolodstep-db — COMMENTED** [review]
Overall looks good, left a comment

**Comment by juliuszsompolski** [comment]
Looks like `loadFromEnvironment()` could be called at Builder creation time, so that it's seeded with the settings from env, and then can be overiden by explicit calls.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 2 above)
Comment (juliuszsompolski) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:934]: :facepalm: I'm blind and didn't notice the `SparkSession.this.create` vs `sessions.get`

**Review by juliuszsompolski — APPROVED** [review]

**Review by vsevolodstep-db — APPROVED** [review]
lgtm

**Review by hvanhovell — APPROVED** [review]
LGTM

**Pull request closed without merging**